src/lib/swapper/swappers/ThorchainSwapper/utils/getL1ToLongtailQuote.ts

```diff
@@ -0,0 +1,174 @@
+import { isErrorPromise, isFulfilled, isRejected } from 'lib/utils'
+import { convertDecimalPercentageToBasisPoints } from 'state/slices/tradeQuoteSlice/utils'
+
+import type { ThorTradeQuote } from '../getThorTradeQuote/getTradeQuote'
+import { addAggregatorAndDestinationToMemo } from './addAggregatorAndDestinationToMemo'
+import { getBestAggregator } from './getBestAggregator'
+import { getL1quote } from './getL1quote'
+import type { AggregatorContract } from './longTailHelpers'
+import { getTokenFromAsset, getWrappedToken, TradeType } from './longTailHelpers'
+
+// This just gets uses UniswapV3 to get the longtail quote for now.
+export const getL1ToLongtailQuote = async (
+  input: GetTradeQuoteInput,
+  streamingInterval: number,
+  assetsById: AssetsByIdPartial,
+): Promise<Result<ThorTradeQuote[], SwapErrorRight>> => {
+  const {
+    buyAsset,
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: sellAmountCryptoBaseUnit,
+    sellAsset,
+    slippageTolerancePercentageDecimal,
+  } = input
+
+  /*
+    We only support L1 -> ethereum longtail swaps for now.
+  */
+  if (buyAsset.chainId !== ethChainId) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getThorTradeQuote] - Unsupported chainId ${buyAsset.chainId}.`,
+        code: TradeQuoteError.UnsupportedChain,
+        details: { buyAssetChainId: buyAsset.chainId },
+      }),
+    )
+  }
+
+  const chainAdapterManager = getChainAdapterManager()
+  const sellChainId = sellAsset.chainId
+  const buyChainId = buyAsset.chainId
+
+  const nativeSellAssetId = chainAdapterManager.get(sellChainId)?.getFeeAssetId()
+  const nativeSellAsset = nativeSellAssetId ? assetsById[nativeSellAssetId] : undefined
+
+  const nativeBuyAssetId = chainAdapterManager.get(buyChainId)?.getFeeAssetId()
+  const nativeBuyAsset = nativeBuyAssetId ? assetsById[nativeBuyAssetId] : undefined
+
+  if (!nativeBuyAsset) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getThorTradeQuote] - No native buy asset found for ${buyChainId}.`,
+        code: TradeQuoteError.InternalError,
+        details: { buyAssetChainId: buyChainId },
+      }),
+    )
+  }
+
+  if (!nativeSellAsset) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getThorTradeQuote] - No native buy asset found for ${sellChainId}.`,
+        code: TradeQuoteError.InternalError,
+        details: { sellAssetChainId: sellChainId },
+      }),
+    )
+  }
+
+  // TODO: Move these constants outside
+  // TODO2: use more than just UniswapV3, and also consider trianglar routes.
+  const ALLOWANCE_CONTRACT = '0xF892Fef9dA200d9E84c9b0647ecFF0F34633aBe8' // TSAggregatorTokenTransferProxy
+
+  const l1Tol1QuoteInput: GetTradeQuoteInput = {
+    ...input,
+    buyAsset: nativeBuyAsset,
+    sellAsset,
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: sellAmountCryptoBaseUnit,
+  }
+
+  const thorchainQuotes = await getL1quote(
+    l1Tol1QuoteInput,
+    streamingInterval,
+    TradeType.L1ToLongTail,
+  )
```

Comment:
This is exactly what I don't like, we are first asking for the L1 quote so we can retrieve the amount of L1 tokens we will definitly have after the thorchain "swap"

Without this, we can't simulate the L1 tokens -> Longtail token swap with the aggregator contract

If you have any other workaround for me to know, I would appreciate it 🙏

Reply:
Yeah that's inherent to the long tail nature - without being able able to calculate the intermediate L1 amount first, we don't know the amount that will be received by the THOR contract to make an AMM swap.

Since THOR is very unlikely to ever expose aggregator quotes in their API (this is a second-class citizen, not a THOR primitive) then that's the way for now and probably forever!